### ISSUE TITLE
About the Conv1DClassifier

### ISSUE THREAD

**Issue by huhk-sysu** [asyml/texar-pytorch#136]
Hello, I'm trying to reduplicate an example (which is originally implemented with `texar`) using `texar-pytorch`. But I got confused about the `Conv1DClassifier`.

1. The first question is about the [documents](https://texar-pytorch.readthedocs.io/en/latest/code/modules.html#texar.torch.modules.Conv1DNetwork):

![image](https://user-images.githubusercontent.com/14769033/62278419-9d534280-b47a-11e9-9fd4-be7859c5bb4e.png)

However there doesn't seem to be a hyperparameter called `filter`. Maybe it means the `out_channels`? 

2. Currently the `Conv1DClassifier` requires an inputs with a shape of `[batch_size, channels, length]`, which means I have to manually transpose the input because in most case it have a shape like `[batch_size, length, dim]`.  I understand that the requirement resulting from pyTorch's `torch.nn.Conv1d`, but I wonder if the classifier can do the transposing job for me so that its  input style may keep in line with other classifiers (always set the `length` as the second dim). 

3. In the Tensorflow version of `texar`, I can set `'other_conv_kwargs': {'padding': 'same'}` to make sure the output shape will be identical to the input shape, but I can't do this in `texar-pytorch` since `torch.nn.Conv1d` only accept an integer for the padding numbers. Now I have 3 different kernel sizes [3, 4, 5] and my workaround is `'other_conv_kwargs': {'padding': 2}`. But I wonder if I can manually set different paddings for different kernel sizes like `padding: [1, 2, 2]`. It would be even better if `texar` can calculate the expected paddings for me.

4. the `Conv1DClassifier` will add a default hyperparameter:

https://github.com/asyml/texar-pytorch/blob/507932c899ca3a8663479b31efc3a41bc7180693/texar/torch/modules/classifiers/conv_classifiers.py#L136-L143

When I set `'num_dense_layers': 0`, The `Linear` layer below will still have a hyperparameter `in_features`, which is totally wrong because there isn't any dense layer in the `Conv1DEncoder`.

https://github.com/asyml/texar-pytorch/blob/507932c899ca3a8663479b31efc3a41bc7180693/texar/torch/modules/classifiers/conv_classifiers.py#L71-L85

In fact in this situation, the correct `in_features` should be the number of different kernel sizes times the number of out channels. Say, I have `'kernel_size': [3, 4, 5]` and `'out_channels': 128`, after Encoder-Flatten-Dropout, the `in_features` should be len([3,4,5]) * 128 = 384

**Comment by huzecong** [comment]
Thanks for the feedback and suggestions! We're looking into the issues and also investigating the possibility of adding support for the features you requested.

**Comment by AvinashBukkittu** [comment]
Thank you for your feedback. After a brief discussion, we have thought of the following changes

1) Yes, you are right, it refers to `out_channels`. We will soon update it in the docs.

2) Even though the norm in Pytorch is `(batch_size, channel_dim, *)`, we will add support for tensor with format `(batch_size, *, channel_dim)`. To achieve this, we will add `data_format` as an argument in ConvNetworks which will default to `channels_first`. If last dimension represents the channel dimension, `data_format` value should be set to `channels_last`.

3) We have decided to make `other_conv_kwargs` accept either a list or a dict. If it is a dict, the same kwargs will be applied to all the conv layers. If it is a list, then `len(other_conv_kwargs) == num_conv_layers` should be true. Additionally, if there are multiple kernel sizes at ith convolution, `other_conv_kwargs[i]` can be a list (of length equal to the corresponding number of kernel sizes) or a dict. If a dict, the same property applies to all the layers at index `i`, otherwise each layer at index `i` gets corresponding kwargs from the list at index `i` [We have also considered the possibility of adding `padding=same` and `padding=valid` options in our framework, but will not be part of this fix].

4) Yes, you are right. Thanks for pointing this out. If `num_dense_layers == 0`, then `in_features` of logits layer should match the output of `Flatten` layer. This will be fixed along with other updates.

Let us know of any questions you may have.

**Comment by huhk-sysu** [comment]
Thanks for your quick response. I think there's nothing else to add since you have a thoughtful consideration. The only question is: Are you talking about the changes in the lateset release (v0.0.1) or in the future version?
I'm grateful for your work which saves me a lot of time. Hoping it become better.

**Comment by AvinashBukkittu** [comment]
The fixes for above will be part of a PR created to solve the issue. Once the PR is merged, you can checkout the master branch. You won't have to wait until a future release.

**Comment by huhk-sysu** [comment]
Sorry to disturb but I found a problem just now.

I noticed you used `time_major=False` here which assums the input will have a shape `[batch_size, max_time, ...]` and masks the second dimension. 

https://github.com/asyml/texar-pytorch/blob/7f3861e4b1b80774a86eeda8fa5cc4029d628c75/texar/torch/modules/networks/conv_networks.py#L523-L525

When I pass a tensor (that I manually transposed)  with a shape `[batch_size, channel_dim, max_time]` to the Conv1D network, it will wrongly mask the sequence.

Also, when the `data_format` is `channels_last`, maybe you need to mask the sequences before permuting the tensor (the latest PR seems to neglect it).

**Comment by AvinashBukkittu** [comment]
I think you are right. This means the error was present in the old code as well because the old code by default dealt with `channel_first` tensors and masking was done using `time_major=False`. Thanks for letting us know.
